libgnucash/engine/ScrubBudget.c

```diff
@@ -0,0 +1,205 @@
+
+    result =
+        heuristics.expense < 0 ? HEURISTICS_INC_EXP :
+        heuristics.liability > 0 ? HEURISTICS_NONE :
+        HEURISTICS_CREDIT_ACC;
+
+    LEAVE ("heuristics_on_budget %s: A(%d) L(%d) Inc(%d) Exp(%d) Eq(%d) = %d",
+           gnc_budget_get_name (budget),
+           heuristics.asset, heuristics.liability, heuristics.income,
+           heuristics.expense, heuristics.equity, result);
+
+    return result;
+}
+
+static void
+fix_budget_acc_sign (Account *acc, gpointer user_data)
+{
+    gnc_numeric amt;
+    ReversalType* reversal = (ReversalType*) user_data;
+    GncBudget* budget = reversal->budget;
+    guint numperiods = gnc_budget_get_num_periods (budget);
+    int type = xaccAccountTypeGetFundamental (xaccAccountGetType (acc));
+
+    ENTER ("budget account reversal [%s] starting", xaccAccountGetName(acc));
+
+    switch (reversal->policy)
+    {
+    case HEURISTICS_INC_EXP:
+        if ((type != ACCT_TYPE_INCOME) && (type != ACCT_TYPE_EXPENSE))
+            return;
+        PINFO ("budget account [%s] is inc/exp. reverse!",
+               xaccAccountGetName(acc));
+        break;
+    case HEURISTICS_CREDIT_ACC:
+        if ((type != ACCT_TYPE_LIABILITY) &&
+            (type != ACCT_TYPE_EQUITY) &&
+            (type != ACCT_TYPE_INCOME))
+            return;
+        PINFO ("budget account [%s] is credit-account. reverse!",
+               xaccAccountGetName(acc));
+        break;
+    case HEURISTICS_NONE:
+        return;
+    }
+
+    for (guint i=0; i < numperiods; ++i)
+    {
+        if (!gnc_budget_is_account_period_value_set (budget, acc, i))
+            continue;
+
+        amt = gnc_budget_get_account_period_value (budget, acc, i);
+        amt = gnc_numeric_neg (amt);
+        gnc_budget_set_account_period_value (budget, acc, i, amt);
+    }
+
+    LEAVE ("budget account reversal [%s] completed!", xaccAccountGetName(acc));
+}
+
+static void
+fix_budget_sign (QofInstance* data, gpointer user_data)
+{
+    GncBudget* budget = GNC_BUDGET(data);
+    Account *root = (Account*) user_data;
+    ReversalType reversal;
+
+    reversal.policy = heuristics_on_budget (budget, root);
+    reversal.budget = budget;
+
+    ENTER ("processing budget [%s] for reversal", gnc_budget_get_name (budget));
+    gnc_account_foreach_descendant (root, fix_budget_acc_sign, &reversal);
+    LEAVE ("completed budget [%s] for reversal", gnc_budget_get_name (budget));
+}
+
+gboolean
+gnc_scrub_budget_signs (QofBook *book)
+{
+    Account* root = gnc_book_get_root_account (book);
+    gchar *retval = NULL;
+
+    if (gnc_features_check_used (book, GNC_FEATURE_BUDGET_UNREVERSED))
+        return FALSE;
+
+    if (!gnc_budget_get_default (book))
+        return FALSE;
+
+    qof_collection_foreach (qof_book_get_collection (book, GNC_ID_BUDGET),
+                            fix_budget_sign, root);
+
+    gnc_features_set_used (book, GNC_FEATURE_BUDGET_UNREVERSED);
+    return TRUE;
```

Comment:
Is it necessarily true that every current budget needs changing? If not this should be conditional on fix_budget_sign actually doing something.

Reply:
I'll rename the function to `check_and_fix_budget_sign`.

Reply:
That doesn't answer the question.

Reply:
Ok not all budgets need changing. If algorithm concludes `HEURISTICS_NONE` then the signs are not reversed, but the feature must still be set to prevent repeating scrubbing.